### PR TITLE
GH-120507: Lower the `BEFORE_WITH` and `BEFORE_ASYNC_WITH` instructions.

### PR DESCRIPTION
**Pull request by markshannon**
Removes the `BEFORE_WITH` and `BEFORE_ASYNC_WITH` instructions, adding the new `LOAD_SPECIAL` instruction.

`BEFORE_WITH` becomes
```
   COPY                     1
   LOAD_SPECIAL             1
   SWAP                     2
   SWAP                     3
   LOAD_SPECIAL             0
   CALL                     0
```
which may seem rather a lot, but it is much more amenable to optimization.

[Negligible performance impact](https://github.com/faster-cpython/benchmarking-public/tree/main/results/bm-20240617-3.14.0a0-68df0a5)

* Issue: gh-120507

### CODE REVIEW

**Review by iritkatriel — APPROVED** [review]

File: Python/optimizer_cases.c.h [hunk 1]
```diff
@@ -1368,9 +1368,16 @@
...
         }
 
-        /* _BEFORE_ASYNC_WITH is not a viable micro-op for tier 2 */
-
-        /* _BEFORE_WITH is not a viable micro-op for tier 2 */
+        case _LOAD_SPECIAL: {
+            _Py_UopsSymbol *attr;
+            _Py_UopsSymbol *self_or_null;
+            attr = sym_new_not_null(ctx);
+            self_or_null = sym_new_not_null(ctx);
```
Comment (TeamSpen210) [inline Python/optimizer_cases.c.h]: This doesn’t seem right, it can push null?

Comment (markshannon) [reply Python/optimizer_cases.c.h]: Yes, the default is `sym_new_not_null` as only one or two instructions push NULL. We'll need to be explicit here.

**Review by Fidget-Spinner — COMMENTED** [review]
Optimizer changes LGTM.

**Comment by markshannon** [comment]
The four JIT/aarach64 failures are a known issue.

File: Python/bytecodes.c [hunk 2]
```diff
@@ -2814,79 +2814,27 @@ dummy_func(
...
-                ERROR_NO_POP();
-            }
-            DECREF_INPUTS();
-            res = PyObject_CallNoArgs(enter);
-            Py_DECREF(enter);
-            if (res == NULL) {
-                Py_DECREF(exit);
-                ERROR_IF(true, error);
             }
+            ERROR_IF(attr == NULL, error);
```
Comment (Fidget-Spinner) [inline Python/bytecodes.c:2828]: @markshannon I think you forgot to `DECREF_INPUTS` here.

Comment (Fidget-Spinner) [reply Python/bytecodes.c:2828]: Oh nevermind the `_PyObject_LookupSpecialMethod` steals it.

**Review by SamyDgz — APPROVED** [review]

**Comment by encukou** [comment]
This added a bunch macros without a Py/_Py prefix to public headers.
I filed #135922 to move them.

**Pull request merged**